### DISCUSSION TITLE
New warning for duplicate case statement

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5499] [Language Ideas]
Imagine this code snippet:
```cs
enum MyEnum { A, B, C };
var x = MyEnum.A;
int res;
switch(x)
{
  case MyEnum.A:
    res = 1;
    break;
  default:
    switch(x)
    {
      case MyEnum.A: // this should become a warning; "unreachable code detected"
        // we know this can't happen because we're in default of outer switch
        res = 1;
        break;
      case MyEnum.B:
        res = 2;
        break;
      case MyEnum.C:
        res = 3;
        break;
    }
    ...
    break;
}
int y = res; // if the inner "case A" is commented, currently this becomes an error - "use of unassigned var"
// it should not be - we know all code paths assign to "res"
```

**Comment by RikkiGibson** [comment]
I think a custom analyzer will be needed to do this. The language/compiler is unlikely to introduce an analysis which tracks which subsets of values may or may not be present in a variable. e.g. we will also probably never introduce a warning for this:

```cs
void M(int x)
{
    if (x == 1)
    {
        if (x == 1)
        {
            // do stuff
        }
    }
}
```

**Reply by yaakov-h** [reply]
This might be something that [CA1508](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508) could cover, or a related analyser.

**Reply by TahirAhmadov** [reply]
@RikkiGibson I think in your example, the inner `if` should have `x != 1`, then it would be CA1508. However, if it is `x == 1` in the inner `if`, like it is verbatim in your post now, it could be a different warning - "performing unnecessary condition check".
A custom analyzer cannot alter variable assignment "flow", and like with other such improvements the purpose of a language-level warning is to prevent everybody from making these mistakes, not just me. I don't need to "self analyze" - I know what to look for, in this case. Others may not notice these.
